pyquil/device.py

```diff
@@ -0,0 +1,315 @@
+        :param Dict[str,Any] d: The dictionary representation.
+        :return: The restored ISA.
+        :rtype: ISA
+        """
+        return cls(
+            name=d["id"]["name"],
+            version=d["id"].get("version", "0.0"),
+            timestamp=d["id"].get("timestamp"),
+            qubits=[Qubit(id=q["qubit-id"], type=q.get("type"), dead=q.get("dead"))
+                    for q in d["logical-hardware"][0]],
+            edges=[Edge(targets=e["action-qubits"], type=e.get("type"), dead=e.get("dead"))
+                   for e in d["logical-hardware"][1]],
+        )
+
+
+_KrausModel = namedtuple("_KrausModel", ["gate", "params", "targets", "kraus_ops", "fidelity"])
+
+
+class KrausModel(_KrausModel):
+    """
+    Encapsulate a single gate's noise model.
+
+    :ivar str gate: The name of the gate
+    :ivar Sequence[float] params: Optional parameters for the gate.
+    :ivar Sequence[int] targets: The target qubit ids.
+    :ivar Sequence[np.array] kraus_ops: The Kraus operators (must be square complex numpy arrays).
+    :ivar float fidelity: The average gate fidelity associated with the Kraus map relative to the
+        ideal operation.
+    """
+
+    @staticmethod
+    def unpack_kraus_matrix(m):
+        """
+        Helper to optionally unpack a JSON compatible representation of a complex Kraus matrix.
+
+        :param Union[list,np.array] m: The representation of a Kraus operator. Either a complex
+            square matrix (as numpy array or nested lists) or a pair of real matrices (as numpy
+            arrays or nested lists) representing the element-wise real and imaginary part of m.
+        :return: A complex square numpy array representing the Kraus operator.
+        :rtype: np.array
+        """
+        m = np.asarray(m, dtype=complex)
+        if m.ndim == 3:
+            m = m[0] + 1j * m[1]
+        if not m.ndim == 2:  # pragma no coverage
+            raise ValueError("Need 2d array.")
+        if not m.shape[0] == m.shape[1]:  # pragma no coverage
+            raise ValueError("Need square matrix.")
+        return m
+
+    def to_dict(self):
+        """
+        Create a dictionary representation of a KrausModel.
+
+        For example::
+
+            {
+                "gate": "RX",
+                "params": np.pi,
+                "targets": [0],
+                "kraus_ops": [            # In this example single Kraus op = ideal RX(pi) gate
+                    [[[0,   0],           # element-wise real part of matrix
+                      [0,   0]],
+                      [[0, -1],           # element-wise imaginary part of matrix
+                      [-1, 0]]]
+                ],
+                "fidelity": 1.0
+            }
+
+        :return: A JSON compatible dictionary representation.
+        :rtype: Dict[str,Any]
+        """
+        res = self._asdict()
+        res['kraus_ops'] = [[k.real.tolist(), k.imag.tolist()] for k in self.kraus_ops]
+        return res
+
+    @classmethod
```

Comment:
Change to staticmethod and use class directly if you don't plan to subclass this class